containertest/README.md

```diff
@@ -0,0 +1,112 @@
+# Mysqltest library
+
+## Introduction
+
+This is a Go library for starting an ephemeral Docker container. It is used to test
+integration between Go code and services such as PostgreSQL.
+
+## How to use it
+
+For some containers such as DBs, startup can take a while. You may want to reuse
+the same container and `USE` separate databases/namespaces or separate tables to
+isolate your tests from each other.
+
+Call `MustStart()` from your `TestMain()` function, If you're not familiar with Go's `TestMain()`
+mechanism for global test initialization, see the docs: https://pkg.go.dev/testing#hdr-Main.
+
+MySQL will be used as an example service, though any Service implementation could
+be used. Currently two implementations of `containertest.Service` exist in this
+repo, stored in `mysql.go` and `postgres.go`.
+
+```go
+package mypackage
+
+import (
+    "database/sql"
+    "fmt"
+    "io"
+    "net"
+    "os"
+    "testing"
+
+    "github.com/abcxyz/pkg/containertest"
+    _ "github.com/go-sql-driver/mysql" // Link with the Go MySQL driver
+)
+
+var ci containertest.ConnInfo
+var mySQLService *containertest.MySQL
+
+// TestMain runs once at startup to do test initialization common to all tests.
+func TestMain(m *testing.M) {
+    var closer io.Closer
+
+	mySQLService := (&MySQL{}).WithVersion("8.0")
```

Comment:
This isn't _super_ common style for Go. Usually it would be:

```go
mySQLService := &MySQL{
  Version: "8.0"
}
```

Reply:
I was copying the existing style: https://github.com/abcxyz/pkg/blob/main/mysqltest/config.go

Is there a reason the pattern was used there but shouldn't be used here? Or would you also change that to use public fields if you had the chance?

Reply:
It's the "options pattern" versus the "struct pattern", but usually the options pattern looks like this:

```go
type Option(a) (a)
func NewThing(ctx context.Context, opts ...Option)
```

Idiomatically, I would expect either:

```go
svc := NewMySQL(ctx, WithVersion("8.0"))
```

or

```go
svc := &MySQL{
  Version: "8.0",
}
```

The "builder pattern" (`foo.WithBar().WithBaz()`) is rarer and usually exists to maintain backwards compatibility with an existing API.

Reply:
Switched to the struct pattern